src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
       .map(l => WithContext(l, TextContext(l, file, None)))
 }
 
+class HeadRDD[T](prev: RDD[T], num: Int, exact: Boolean)(implicit tct: ClassTag[T]) extends RDD[T](prev) {
+
+  private var nLastPart = 0
+  private var lastPart = 0
+
+  def getPartitions: Array[Partition] = {
+    val buf = new mutable.ArrayBuffer[Int]
+
+    buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length, Seq(0))
+    if (exact || buf.length == 0) {
+      while (buf.sum < num && buf.length < prev.partitions.length) {
+        buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length,
+          buf.length to
+            buf.length +
+              (if (buf.sum > 0) math.min(((num - buf.sum) / (buf.sum / buf.length).toDouble).ceil.toInt, prev.partitions.length)
+              else 1)
+        )
+      }
+
+      def f(n: Int, i: Int): Unit = {
+        if (n + buf(i) > num) {
+          lastPart = i
+          nLastPart = num - n
+        } else {
+          f(n + buf(i), i + 1)
+        }
+      }
+
+      f(0, 0)
+      firstParent.partitions
```

Comment:
I think you can do the filter, and compute the partitions along with setting lastPart and nLastPart all together in `f`.  It reads awkwardly to go over partitions twice right together using two different idioms (recursion vs combinators).